### ISSUE TITLE
advance, distance, next, and prev should be customization point objects

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#367]
Imagine passing a new-style RandomAccessIterator `I` for which `reference_t<I>` is not a reference type to (unqualified) `advance`. Further imagine that `I` has namespace `std` as an associated namespace (by wrapping a `std` iterator, for instance). Then `std::advance` might get called, which will treat the iterator as an InputIterator (due to category demotion), doing *N* increments instead of just one `+=`.

Making `ranges::advance` a CPO could help in some scenarios. Ditto for `distance`, `next`, and `prev`.

Proposed Resolution
-----------------------
[_Editor's note_: this wording assumes that the resolution of #256 has been applied. It depicts distinct overloads for `next(i)` and `next(i, n)` instead of the semantically equivalent `next(i, n = 1)` to allow implementations that strengthen `noexcept` to conformingly reflect `noexcept(++i)` for the `next(i)` case. Otherwise, the "new" wording is as close as possible to the "old" wording to minimize review costs: it simply turns function template overload sets into member function template overload sets.]

Change the synopsis of `<experimental/ranges/iterator>` in [iterator.synopsis] as follows:

```diff
 // 9.6.4, iterator operations:
-template <Iterator I>
-  void advance(I& i, difference_type_t<I> n);
-template <Iterator I, Sentinel<I> S>
-  void advance(I& i, S bound);
-template <Iterator I, Sentinel<I> S>
-  difference_type_t<I> advance(I& i, difference_type_t<I> n, S bound);
-template <Iterator I, Sentinel<I> S>
-  difference_type_t<I> distance(I first, S last);
-template <Iterator I>
-  I next(I x, difference_type_t<I> n = 1);
-template <Iterator I, Sentinel<I> S>
-  I next(I x, S bound);
-template <Iterator I, Sentinel<I> S>
-  I next(I x, difference_type_t<I> n, S bound);
-template <BidirectionalIterator I>
-  I prev(I x, difference_type_t<I> n = 1);
-template <BidirectionalIterator I>
-  I prev(I x, difference_type_t<I> n, I bound);
+namespace {
+  constexpr unspecified advance = unspecified;
+  constexpr unspecified distance = unspecified;
+  constexpr unspecified next = unspecified;
+  constexpr unspecified prev = unspecified;
+}

 [...]

 template <Iterator I>
   counted_iterator<I> make_counted_iterator(I i, difference_type_t<I> n);

-template <Iterator I>
-   void advance(counted_iterator<I>& i, difference_type_t<I> n);

 // 9.7.8, unreachable sentinels:

 [...]

 // 9.11, range primitives:
 namespace {
   constexpr unspecified size = unspecified ;
   constexpr unspecified empty = unspecified ;
   constexpr unspecified data = unspecified ;
   constexpr unspecified cdata = unspecified ;
 }
-template <Range R>
-difference_type_t<iterator_t<R>> distance(R&& r);
-template <SizedRange R>
-difference_type_t<iterator_t<R>> distance(R&& r);
```

In [iterator.operations], change paragraph 1 as follows:
```diff
 1 Since only types that satisfy RandomAccessIterator provide the + operator, and types that
  satisfy SizedSentinel provide the - operator, the library provides four
- function templates
+ customization point objects
  advance, distance, next, and prev. These
- function templates
+ customization point objects
  use + and - for random access iterators and ranges that satisfy SizedSentinel
- , respectively
  (and are, therefore, constant time for them); for output, input, forward and bidirectional
  iterators they use ++ to provide linear time implementations.
```

and replace the remainder of [iterator.operations] entirely with:

> 2 The name `advance` denotes a customization point object (\ref{customization.point.object}). It has the following function call operators:
> ```c++
> template <Iterator I>
>   constexpr void operator()(I& i, difference_type_t<I> n) const;
> ```
> 3 *Requires:* `n` shall be negative only for bidirectional iterators.
>
> 4 *Effects:* For random access iterators, equivalent to `i += n`. Otherwise, increments (or decrements for negative `n`) iterator `i` by `n`.
> ```c++
> template <Iterator I, Sentinel<I> S>
>   constexpr void operator()(I& i, S bound) const;
> ```
> 5 *Requires:* If `Assignable<I&, S>()` is not satisfied, `[i,bound)` shall denote a range.
>
> 6 *Effects:*
>
> (6.1) — If `Assignable<I&, S>()` is satisfied, equivalent to `i = std::move(bound)`.
>
> (6.2) — Otherwise, if `SizedSentinel<S, I>()` is satisfied, equivalent to `advance(i, bound - i)`.
>
> (6.3) — Otherwise, increments `i` until `i == bound`.
> ```c++
> template <Iterator I, Sentinel<I> S>
>   constexpr difference_type_t<I> operator()(I& i, difference_type_t<I> n, S bound) const;
> ```
> 7 *Requires:* If `n > 0`, `[i,bound)` shall denote a range. If `n == 0`, `[i,bound)` or `[bound,i)` shall denote a range. If `n < 0`, `[bound,i)` shall denote a range and `(BidirectionalIterator<I>() && Same<I, S>())` shall be satisfied.
>
> 8 *Effects:*
>
> (8.1) — If `SizedSentinel<S, I>()` is satisfied:
>
> (8.1.1) — If `|n| >= |bound - i|`, equivalent to `advance(i, bound)`.
>
> (8.1.2) — Otherwise, equivalent to `advance(i, n)`.
>
> (8.2) — Otherwise, increments (or decrements for negative `n`) iterator `i` either `n` times or until `i == bound`, whichever comes first.
>
> 9 *Returns:* `n - M`, where `M` is the distance from the starting position of `i` to the ending position.
>
> 10 The name `distance` denotes a customization point object. It has the following function call operators:
> ```c++
> template <Iterator I, Sentinel<I> S>
>   constexpr difference_type_t<I> operator()(I first, S last) const;
> ```
> 9 *Requires:* `[first,last)` shall denote a range, or `(Same<S, I>() && SizedSentinel<S, I>())` shall be satisfied and `[last,first)` shall denote a range.
>
> 10 *Effects:* If `SizedSentinel<S, I>()` is satisfied, returns `(last - first)`; otherwise, returns the number of increments needed to get from `first` to `last`.
>
> ```c++
> template <Range R>
>   constexpr difference_type_t<iterator_t<R>> operator()(R&& r) const;
> ```
> 11 *Effects:* Equivalent to: `return distance(ranges::begin(r), ranges::end(r));` (\ref{iterator.range})
>
> [_Editor's note:_ Include the following paragraph only if the PR of #211 has been applied.]
>
> 12 *Remarks:* Instantiations of this function template may be ill-formed if the declarations in `<experimental/range/range>` are not in scope at the point of instantiation (\cxxref{temp.point}).
>
> ```c++
> template <SizedRange R>
>   constexpr difference_type_t<iterator_t<R>> operator()(R&& r) const;
> ```
> 13 *Effects:* Equivalent to: `return ranges::size(r);` (\ref{range.primitives.size})
>
> [_Editor's note:_ Include the following paragraph only if the PR of #211 has been applied.]
>
> 14 *Remarks:* Instantiations of this function template may be ill-formed if the declarations in `<experimental/range/range>` are not in scope at the point of instantiation (\cxxref{temp.point}).
>
> 15 The name `next` denotes a customization point object. It has the following function call operators:
> ```c++
> template <Iterator I>
>   constexpr I operator()(I x) const;
> ```
> 16 *Effects:* Equivalent to: `++x; return x;`
> ```c++
> template <Iterator I>
>   constexpr I operator()(I x, difference_type_t<I> n) const;
> ```
> 17 *Effects:* Equivalent to: `advance(x, n); return x;`
> ```c++
> template <Iterator I, Sentinel<I> S>
>   constexpr I operator()(I x, S bound) const;
> ```
> 18 *Effects:* Equivalent to: `advance(x, bound); return x;`
> ```c++
> template <Iterator I, Sentinel<I> S>
>   constexpr I operator()(I x, difference_type_t<I> n, S bound) const;
> ```
> 19 *Effects:* Equivalent to: `advance(x, n, bound); return x;`
>
> 20 The name `prev` denotes a customization point object. It has the following function call operators:
> ```c++
> template <BidirectionalIterator I>
>   constexpr I operator()(I x) const;
> ```
> 21 *Effects:* Equivalent to: `--x; return x;`
> ```c++
> template <BidirectionalIterator I>
>   constexpr I operator()(I x, difference_type_t<I> n) const;
> ```
> 22 *Effects:* Equivalent to: `advance(x, -n); return x;`
> ```c++
> template <BidirectionalIterator I, Sentinel<I> S>
>   constexpr I operator()(I x, difference_type_t<I> n, S bound) const;
> ```
> 23 *Effects:* Equivalent to: `advance(x, -n, bound); return x;`

Also strike the declaration of the `advance` overload from the `counted_iterator` synopsis in [counted.iterator], and strike its specification: paragraphs 9 and 10 in [counted.iter.nonmember].

Strike paragraphs 1 & 2 from [range.primitives]:
```diff
-template <Range R>
-difference_type_t<iterator_t<R>> distance(R&& r);
-
-1 Effects: Equivalent to: return ranges::distance(ranges::begin(r), ranges::end(r));
-
-template <SizedRange R>
-difference_type_t<iterator_t<R>> distance(R&& r);
-
-2 Effects: Equivalent to: return ranges::size(r);
```

**Comment by CaseyCarter** [comment]
Considering caseycarter/cmcstl2#12, our guidance should be to *always* call components in the Ranges library with qualified names.

(This is not opposition to the suggestion, just pointing out that unqualified names have many dangers.)

**Comment by CaseyCarter** [comment]
For clarification: are you suggesting that these names be actual customization points, or simply that they be CPOs to kill ADL? `counted_iterator` witnesses that at least `advance` could usefully be a true customization point.

**Comment by ericniebler** [comment]
> For clarification: are you suggesting that these names be actual customization points, or simply that they be CPOs to kill ADL? 

Mostly just to kill ADL.

> `counted_iterator` witnesses that at least `advance` could usefully be a true customization point.

I'm ambivalent on `advance` being a CPO. The conservative thing would be to just hard-code the optimization for `counted_iterator` and consider making `advance` a true CPO separately.

**Comment by CaseyCarter** [comment]
Note that this issue conflicts somewhat with #211. If `distance` becomes a CPO, we cannot relocate range overloads of `distance` into `<experimental/range/range>`. It seems reasonable to define the CPO in `iterator`, ensure that `range` makes it available, and allow calls to `distance(range)` to be ill-formed if `range` hasn't been included.

I'll write up the PR with wording for the "no #211" case, and the "#211 as proposed" case. We'll simply have to fix this issue if #211 changes significantly.

**Comment by CaseyCarter** [comment]
@ericniebler PR in the OP for your inspection. It is large enough that I have made an effort to preserve the form of the original wording to ease review and reduce the potential for LWG-required edits.
